mfkdf2/src/derive/factors/totp.rs

```diff
@@ -104,53 +81,46 @@ impl FactorDerive for TOTP {
   fn params(&self, key: Key) -> MFKDF2Result<Self::Params> {
-    let params: Value = serde_json::from_str(&self.params)?;
+    let params: TOTPParams = serde_json::from_value(self.params.clone())?;
 
-    let pad = params["pad"].as_str().ok_or(MFKDF2Error::MissingDeriveParams("pad".to_string()))?;
-    let pad = base64::prelude::BASE64_STANDARD.decode(pad)?;
+    let pad = base64::prelude::BASE64_STANDARD.decode(params.pad)?;
     let padded_secret = decrypt(pad.clone(), &key.0);
 
-    let window =
-      params["window"].as_u64().ok_or(MFKDF2Error::MissingDeriveParams("window".to_string()))?;
-    let step =
-      params["step"].as_u64().ok_or(MFKDF2Error::MissingDeriveParams("step".to_string()))?;
-    let digits =
-      params["digits"].as_u64().ok_or(MFKDF2Error::MissingDeriveParams("digits".to_string()))?;
-    let hash: HashAlgorithm = serde_json::from_value(params["hash"].clone())?;
-
-    let time =
-      self.options.time.ok_or(MFKDF2Error::MissingDeriveParams("time".to_string()))? as u128;
-    let mut new_offsets = Vec::with_capacity((4 * window) as usize);
+    let time = params.start;
+    let mut new_offsets = Vec::with_capacity((4 * params.window) as usize);
 
-    for i in 0..window {
-      let counter = (time / 1000) as u64 / step + i;
-      let code = generate_hotp_code(&padded_secret[..20], counter, &hash, digits as u8);
+    for i in 0..params.window {
+      let counter = (time / 1000) as u64 / params.step + i;
+      let code =
+        generate_hotp_code(&padded_secret[..20], counter, &params.hash, params.digits as u8);
 
       let mut offset =
-        mod_positive(self.target as i64 - code as i64, 10_i64.pow(digits as u32)) as u32;
+        mod_positive(self.target as i64 - code as i64, 10_i64.pow(params.digits as u32)) as u32;
 
-      let oracle_time = counter * step * 1000;
+      let oracle_time = counter * params.step * 1000;
       if self.options.oracle.is_some()
         && self.options.oracle.as_ref().unwrap().contains_key(&oracle_time)
       {
         offset = mod_positive(
           offset as i64 + *self.options.oracle.as_ref().unwrap().get(&oracle_time).unwrap() as i64,
-          10_i64.pow(digits as u32),
+          10_i64.pow(params.digits as u32),
         ) as u32;
       }
 
       new_offsets.extend_from_slice(&offset.to_be_bytes());
     }
 
-    Ok(json!({
-      "start": time,
-      "hash": hash.to_string(),
-      "digits": digits,
-      "step": step,
-      "window": window,
-      "pad": base64::prelude::BASE64_STANDARD.encode(&pad),
-      "offsets": base64::prelude::BASE64_STANDARD.encode(&new_offsets),
-    }))
+    let params = TOTPParams {
+      start:   time,
+      hash:    params.hash,
+      digits:  params.digits,
+      step:    params.step,
+      window:  params.window,
+      pad:     base64::prelude::BASE64_STANDARD.encode(&pad),
+      offsets: base64::prelude::BASE64_STANDARD.encode(&new_offsets),
+    };
+
+    Ok(serde_json::to_value(params)?)
```

Comment:
Seems like we could just make the return type `TOTPParams` and enforce the trait to have the `type Params: Serialize + for<'de>Deserialize<'de>` so that serialization only happens when we need to pass this out of Rust most likely. What do you think? (I think this was our goal initially too)

Reply:
I think this is the one #39